README.md
![geniem-github-banner](https://cloud.githubusercontent.com/assets/5691777/14319886/9ae46166-fc1b-11e5-9630-d60aa3dc4f9e.png)
# DustPress Plugin: DustPress.js

A DustPress plugin that provides a handy JavaScript library for using your DustPress model methods on the front end.

- Contributors: [devgeniem](https://github.com/devgeniem) / [Nomafin](https://github.com/Nomafin), [villesiltala](https://github.com/villesiltala), [godbone](https://github.com/godbone)
- Plugin url: https://github.com/devgeniem/dustpress-debugger
- Tags: dustpress, wordpress, plugins, dustjs, dust.js
- Requires at least: 4.9.0
- Requires DustPress version: 1.25.0
- Tested up to: 5.2.0
- License: GPL-3.0
- License URI: http://www.gnu.org/licenses/gpl-3.0.html

## Usage

You can call for `SomeModel`'s method `SomeMethod` with following code:

```js
dp( 'SomeModel/SomeMethod', {
    tidy: true,
    args: {
        'foo': 'bar'
    }
}).then( ( data ) => {
    // do what you want with the data
}).catch( ( error ) => {
    // possible error
});
```

`tidy: true` parameter cleans up the data tree a bit for more usability. Feel free to try the queries with and without it to see the difference.

`args` parameter passes arguments to the PHP side. They can be accessed there with `$this->get_args();`.

If you want, you can even render HTML with Dust templates.

```js
dp( 'SomeModel/SomeMethod', {
    partial: 'SomePartial',
}).then( ( data ) => {
    // do what you want with the data
}).catch( ( error ) => {
    // possible error
});
```
This code takes the data of `SomeMethod` and renders it with `SomePartial`. Variable `data` then contains the ready html.

If you still want to get the data output as well, use argument `data: true` and you will get the resulting data as the second parameter of your success function.

You can also omit the method completely if you want to get the data of a complete model.

```js
dp( 'SomeModel/SomeMethod' ).then( ( data ) => {
    // do what you want with the data
}).catch( ( error ) => {
    // possible error
});
```

It is also possible to use the `dp` call with the async-await pattern:

```js
try {
    var foobar = await dp( 'SomeModel/SomeMethod', {
        tidy: true,
        args: {
            foo: 'bar'
        }
    });
} catch( err ) {
    console.error( err );
}
```

Now data will consist of an object with the methods' names as keys and their return values as the values. Obviously you can also render that to HTML as well.

### Additional parameters

#### bypassMainQuery

By default DustPress.js requests bypass WordPress' main WP_Query so that it wouldn't slow the request down when it's not necessary. You can prevent that from happening by setting `bypassMainQuery: false` if you want to use the default query.

### Model function front-end visibility

You need to make your methods accessible for DustPress.js by defining a property named `$api` in your model. It should be an array consisting of the names of your accessible methods. DustPress.js can also run `protected` methods which are not automatically run by DustPress normally.

```
class SomeModel extends DustPressModel {
    public $api = [
        'SomeMethod'
    ];

    protected function SomeMethod() {
        // Some code..
    }
}
```

If you need to determine whether a call has been made from ajax or not, you can use `dustpress()->is_dustpress_ajax()` function like shown below:

```
class SomeModel extends DustPressModel {
    public $api = [
        'PublicMethod'
    ];

    public function PublicMethod() {
        if ( dustpress()->is_dustpress_ajax() ) {
            // Do not run if this an ajax request.
            return;
        }
        // Some code..
    }
}
```

## Install

Recommended installation to a WordPress project is through composer:
```
$ composer require devgeniem/dustpress-js
```

Obviously you can also download the ZIP file from GitHub and extract it into your plugins directory.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [4.4.1] - 2021-01-15

### Fixed
- Removed an unnecessary and unused filter that caused warnings with PHP 8.0.

## [4.4.0] - 2020-07-02

### Added
- Retry fetch requests on network error by default 3 times

### Fixed
- Get request handling

## [4.3.0] - 2020-03-30

### Added
- Filter for dustpress-js wp_register_script() in footer value.
- Note this is also available in the release 3.1.0 for the older 3.x.x versions.

## [4.1.1] - 2019-11-11

### Fixed

- Update plugin version texts.

## [4.1.0] - 2019-11-11

### Added

- Add back the data property for fetching data along with the rendered HTML.

### Changed

- NPM update.

## [4.0.1] - 2019-11-08

### Added
- DustPress Debugger integration

## [4.0.0] - 2019-10-25

### Added
- webpack compiling for dustpress.js
- js sourcemaps
- .eslintrc.json

### Changed
- Changed the format of dustpress.js to a class
- Use fetch instead of jquery
- Use same CSRF token for all requests in the same session

## [3.0.5] - 2019-01-30

### Added

- Run DustPress debugger update in a new execution queue to prevent it halting resolving the AJAX request.

## [3.0.4] - 2018-08-09

### Fixed
- Fixed an error that can happen if the server doesn't respond anything

## [3.0.3] - 2018-08-08

### Fixed
- Changed the deprecated $.ajax.complete to $.ajax.always

## [3.0.2] - 2018-06-27

### Fixed
- A problem that was caused by the previous fix.

## [3.0.1] - 2018-06-27

### Fixed
- A problem that prevented partial rendering with DustPress.js.

## [3.0.0] - 2018-06-20

### Changed
- `bypassMainQuery` default value as `true`.

## [2.3.0] - 2018-06-20

### Added
- `bypassMainQuery` parameter to be able to bypass main WP_Query if it's not needed on the request.

## [2.2.0] - 2018-06-14

### Added
- Possibility to change the ajax endpoint url either on per-request basis or project-wise.

### Changed
- Default endpoint for the ajax call to be the home url instead of current page url.

## [2.1.2] - 2018-05-25

### Fixed
- A bug where two simultaneous DustPress.js calls would clash and cause unexpected behaviour.

## [2.1.1] - 2018-01-17

### Fixed
- A bug regarding the enqueueing of the script

## [2.1.0] - 2018-01-17

### Added
- Added a feature for getting both the data and the rendered output when using a Dust partial
- Added possibility to change the dependency setting of the DustPress.js library if, for example, jQuery is provided outside the WordPress enqueue system

## [2.0.2] - 2017-11-15

### Changed
- Fixed a bug that caused crashing in certain circumstances in the front-end
- Enhanced the documentation with better examples and fixed some typos

## [2.0.1] - 2017-11-10

### Changed
- Changed the enqueue call to use the plugin version for the script version

## [2.0.0] - 2017-10-20

### Changed
- Changed to using application/json so parameters keep their type. NOTE: this will break earlier functionality as dustpress model `get_args()` function returned an array but it will now return an object instead

## [1.1.3] - 2017-10-04

### Added
- Added a return to the default dp function so the ajax call can be handled
- Added a changelog


js/DPRequest.js
export default class DPRequest {

    /**
     * Request path
     *
     * @type {String}
     */
    path = '';

    /**
     * Default instance params
     *
     * @type {Object}
     */
    params = {
        get: '',
        type: 'post',
        tidy: false,
        render: false,
        partial: null,
        upload: false,
        data: false,
        url: null,
        bypassMainQuery: true,
        contentType: 'application/json',
        token: '',
        success: () => {},
        error: () => {},
        uploadProgress: () => {},
        downloadProgress: () => {},
        start: () => {},
        complete: () => {}
    };

    /**
     * Construct the dustpress request
     *
     * @param {Object} params Params to use in request
     */
    constructor( path, params = {}) {
        this.path = path;

        // Override default params with given ones
        this.params = Object.assign( this.params, params );
    }

    send() {
        if ( this.params.get.length && ! this.params.get.startsWith( '?' ) ) {
            this.params.get = '?' + this.params.get;
        }

        const date = new Date();
        date.setTime( date.getTime() + ( 24 * 60 * 60 * 1000 ) );

        // Set the cookie for the token
        document.cookie = 'dpjs_token=' + this.params.token + '; expires=' + date.toGMTString() + '; path=/';

        const options = {
            url: ( this.params.url || dustpressjs_endpoint || ( window.location + this.params.get ) ),
            method: this.params.type,
            contentType: this.params.contentType,
            data: {
                dustpress_data: {
                    path: this.path,
                    args: this.params.args,
                    render: this.params.render,
                    tidy: this.params.tidy,
                    partial: this.params.partial,
                    data: this.params.data,
                    token: this.params.token,
                    bypassMainQuery: this.params.bypassMainQuery
                }
            }
        };

        // Stringify data so it can be sent
        options.data = JSON.stringify( options.data );

        if ( this.params.upload ) {
            options.xhr = () => this.xhrUpload();
        }

        this.params.start();

        this.params.successHandler          = ( data, textStatus, jqXHR ) => this.successHandler( data, textStatus, jqXHR );
        this.params.errorHandler            = ( jqXHR, textStatus, errorThrown ) => this.errorHandler( jqXHR, textStatus, errorThrown );
        this.params.uploadProgressHandler   = ( event ) => this.uploadDownloadProgressHandler( event );
        this.params.downloadProgressHandler = ( event ) => this.uploadDownloadProgressHandler( event );

        return $.ajax( options )
        .done( ( data, textStatus, jqXHR ) => this.params.successHandler( data, textStatus, jqXHR ) )
        .fail( ( jqXHR, textStatus, errorThrown ) => this.params.errorHandler( jqXHR, textStatus, errorThrown ) )
        .always( ( dataOrXhr, textStatus, xhrOrError ) => this.params.complete( dataOrXhr, textStatus, xhrOrError ) );
    }

    /**
     * Handle ajax success
     *
     * @param  {string|mixed} data       Data retrieved by ajax call.
     * @param  {string}       textStatus Response status.
     * @param  {jqXHR}        jqXHR      Superset of the browser's native XMLHttpRequest object.
     */
    successHandler( data, textStatus, jqXHR ) {
        let parsed;

        if ( typeof data === 'string' ) {
            parsed = $.parseJSON( data );
        } else {
            parsed = data;
        }

        // Expire CSRF cookie
        document.cookie = 'dpjs_token=; expires=-1; path=/';

        // Add to debugger data if it exists
        if ( window.DustPressDebugger ) {
            this.addToDebugger( parsed );
        }

        if ( parsed && ! parsed.error ) {
            this.params.success( parsed.success, parsed.data, textStatus, jqXHR );
        } else {
            this.params.error( parsed, textStatus, jqXHR );
        }
    }

    /**
     * Handle file uploads
     */
    xhrUpload() {
        const xhr = new window.XMLHttpRequest();

        xhr.upload.addEventListener( 'progress', ( event ) => this.params.uploadProgressHandler( event ), false );
        xhr.addEventListener( 'progress', ( event ) => this.params.uploadDownloadProgressHandler( event ), false );
    }

    /**
     * Handle errors.
     *
     * @param  {jqXHR}  jqXHR       Superset of the browser's native XMLHttpRequest object.
     * @param  {string} textStatus  Response status.
     * @param  {mixed}  errorThrown Ajax call error.
     */
    errorHandler( jqXHR, textStatus, errorThrown ) {

        // Expire CSRF cookie
        document.cookie = 'dpjs_token=; expires=-1; path=/';

        this.params.error({error: errorThrown}, textStatus, jqXHR );
    }

    /**
     * Handle uploar or download progress.
     *
     * @param {object} event Event data.
     */
    uploadDownloadProgressHandler( event ) {
        if ( event.lengthComputable ) {
            const complete = ( event.loaded / event.total );

            this.params.downloadProgress( complete );
        }
    }

    /**
     * Add called data to dustpress debugger
     *
     * @param {object} parsed Parsed data.
     */
    addToDebugger( parsed ) {
        delete this.params.params.success;
        delete this.params.params.error;

        const debug = {
            params: this.params.params,
            data: parsed.debug ? parsed.debug : parsed
        };
        window.DustPressDebugger.extend( debug, this.params.path );
    }
}


js/dustpress.js
import 'whatwg-fetch'; // fetch polyfill
import 'core-js/fn/promise'; // Promise polyfill
import Cookie from 'js-cookie';
import fetchRetry from 'fetch-retry';
import jqueryParam from 'jquery-param';

export default class DustPress {

    /**
     * CSRF token
     *
     * @static
     * @type {String}
     */
    static csrfToken = '';

    /**
     * Default dpjs_token cookie params
     *
     * @type {Object}
     */
    cookieParams = {
        expires: 60 * 60 * 1000,
        path: '/'
    }

    /**
     * Default dp() params
     *
     * @type {Object}
     */
    params = {
        url: window.location,
        method: 'POST',
        args: {},
        credentials: 'include',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        bypassMainQuery: false,
        partial: '',
        render: false,
        tidy: false,
        data: false,
        retries: 3,
        retryDelay: 0,
        redirect: 'error',
    }

    /**
     * Generate token for CSRF protection
     *
     * @return {String} Generated token.
     */
    generateToken() {
        // Use the already generated token if we have one
        if ( this.csrfToken ) {
            return this.csrfToken;
        }

        let token = '';

        if ( typeof window.crypto !== 'undefined' && typeof window.crypto.getRandomValues === 'function' ) {
            for ( let i = 0; i < 4; i++ ) {
                token += window.crypto.getRandomValues( new Uint32Array( 1 ) );
            }
        } else {
            token = Math.random() + '' + Math.random();
        }

        this.csrfToken = token;
        Cookie.set( 'dpjs_token', token, this.cookieParams );

        return token;
    }

    /**
     * Call dustpress function
     *
     * @param  {String} path   Dustpress call path.
     * @param  {Object} params Call args.
     * @return {Promise}       Result promise.
     */
    send( path, params = {}) {

        // Get the CSRF token
        const token = this.generateToken();

        // Extract params from the combination of default and given params
        const {
            method,          // Ajax call method
            args,            // Dustpress ajax args
            bypassMainQuery, // Whether the main wp query should be disabled
            partial,         // What partial to render
            render,          // Whether we whould render a partial or not
            tidy,            // Should we tidy the ajax call output
            data,            // If set to true, the response also contains the data used when rendering a template.
            credentials,     // Whether we should send cookie data with the call or not
            headers,         // What headers to use for the call
            retries,         // How mny times to retry failed request on network error
            retryDelay,      // How much time to wait between retries
            redirect         // What should fetch do on redirect
        } = Object.assign({}, this.params, params );
        const url = new URL( params.url || this.params.url );

        // Edge doesnt implement the "finally" function in its fetch but does in a Promise so we need to wrap this in a Promise
        return new Promise( ( resolve, reject ) => {
            const dustpress_data = {
                args,
                bypassMainQuery,
                partial,
                path,
                render,
                tidy,
                data,
                token
            };
            const fetchParams = {
                method,
                credentials,
                headers,
                retries,
                retryDelay,
                redirect
            };

            // If doing a GET request add the params to the url
            if ( method === 'GET' ) {
                dustpress_data.dustpress_data = true;
                url.search = jqueryParam( dustpress_data ).split( '=&' ).join( '&' );
            }
            else {
                fetchParams.body = JSON.stringify({ dustpress_data });
            }

            // Send the request
            fetchRetry( fetch )( url, fetchParams ).then( ( response ) => {
                if ( response.ok ) {
                    return response.json();
                }

                // Reject promise as well if the response was not 200
                throw Error( response.statusText );
            }).then( ( data ) => {
                resolve( data );

                if ( typeof window.DustPressDebugger !== 'undefined' ) {
                    setTimeout( () => {
                        window.DustPressDebugger.extend({
                            params: args,
                            data: data
                        }, path );
                    }, 0 );
                }
            }).catch( ( err ) => {
                reject( err );
            });
        });
    }
}

// Setup global dp() function
if ( typeof window.dp === 'undefined' ) {
    const dustpress = new DustPress();
    window.dp = ( path, params ) => dustpress.send( path, params );
}


webpack.config.js
const webpack = require( 'webpack' );

// Provide jQuery instance for all modules.
const providePlugin = new webpack.ProvidePlugin({
    jQuery: 'jquery',
    '$': 'jquery'
});

/**
 * Webpack options
 * @type {Object}
 *       entry: a file specifying what files to require
 *       output: a folder that speicifes where all compiled output goes and a file where compiled javascript goes
 *       module.rules: loaders to be used for compiling the files, current compilers handle scss and markdown
 */
module.exports = {
    devtool: 'source-map',
    entry: {
        dustpress: __dirname + '/js/dustpress.js'
    },
    output: {
        path: __dirname + '/js/',
        filename: '[name]-min.js'
    },
    module: {
        rules: [
            {
                test: /\.js$/,
                exclude: /node_modules/,
                use: {
                    loader: 'babel-loader',
                    options: {
                        presets: [ '@babel/preset-env' ],
                        plugins: [ '@babel/plugin-proposal-class-properties' ]
                    }
                }
            }
        ]
    },
    plugins: [
        providePlugin
    ],
    externals: {
        'jquery': 'jQuery'
    },
    watchOptions: {
        poll: 100
    }
};


plugin.php
<?php
/**
 * Plugin Name: DustPress.JS
 * Plugin URI: https://github.com/devgeniem/dustpress-js
 * Description: DustPress JavaScript library. Provides a front-end interface for interacting with the backend functions.
 * Version: 4.4.1
 * Author: Geniem Oy / Miika Arponen, Ville Siltala & Arttu Mäkipörhölä
 * Author URI: http://www.geniem.com
 */
class DustPressJs {

    /**
     * Initalizes the plugin
     *
     * @type    function
     * @date    24/05/2016
     * @since   0.1.0
     *
     * @return  void
     */
    public function __construct() {
        add_action( 'init', [ $this, 'enqueue_scripts' ] );
    }

    /**
     *  This function enqueues front end scripts.
     *
     *  @type    function
     *  @date    17/12/2015
     *  @since   0.1.0
     *
     *  @return  void
     */
    public function enqueue_scripts() {

        // Variables.
        $plugin_data  = get_file_data( __FILE__, [ 'Version' => 'Version' ], 'plugin' );
        $version      = $plugin_data['Version'];
        $dependencies = apply_filters( 'dustpress/js/dependencies', [ 'jquery' ] );
        $in_footer    = apply_filters( 'dustpress/js/in_footer', false );

        wp_register_script( 'dustpress', trailingslashit( plugin_dir_url( __FILE__ ) ) . 'js/dustpress-min.js', $dependencies, $version, $in_footer );

        wp_enqueue_script( 'dustpress' );
    }
}

new DustPressJs();
